keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+- Security: provide a solution that enables the secure copy of cluster certificates between
+  control-plane nodes
+
+## Proposal
+
+### Constraints
+
+The solution described in this proposal is deeply influenced by following constraints:
+
+1. Kubeadm can execute actions only on the machine where it is running e.g. it is not
+   possible to execute actions on other nodes.
+2. `kubeadm init` and `kubeadm join` are separated actions (executed on separated machines/at
+   different times).
+3. During the join workflow, kubeadm can access the cluster only using identities with
+   limited grants, namely `system:unauthenticated` or `system:node-bootstrapper`.
+
+### Key element of the proposal
+
+At a high level, the proposed solution can be summarized by the following workflow:
+
+**On control plane node 1:**
+
+```bash
+kubeadm init --upload-certs
+```
+
+The new `--upload-certs` flag will trigger a new kubeadm init phase executing following
+actions:
+
+1. An encryption key will be generated; such **key will never be stored on cluster**
+
+2. Cluster certificates will be encrypted using the above key
+
+3. Encrypted certificates will be stored in a new Kubernetes secret named `kubeadm-certs`;
+   it is important to notice that:
+
+   - This secret is the technical solution that provides a temporary bridge between
+     `kubeadm init` and `kubeadm join` commands/between different nodes.
+
+   - Without the encryption key, this secret contains a harmless blog of bytes.
+
+4. A second bootstrap token will be generated, with a shorter duration than the
+   join token (the TTL-token)
+
+5. The `ownerReference` field in the `kubeadm-certs` secret will be linked to the
+   TTL token, thus ensuring automatic deletion of the secret when the TTL-token gets
+   deleted by the `TokenCleaner` controller
+
+6. RBAC rules will be created to ensure access to the above config map to users
+   in the `system:node-bootstrapper` group (the bootstrap tokens).
+
+**On control plane node 2:**
+
+```bash
+kubeadm kubeadm join --control-plane --certificate-key={key from step above}
+```
+
+The new `--certificate-key` will trigger following actions:
+
+1. The `kubeadm-certs` config will be retrived from the cluster
+2. Cluster certificates will be decrypted using the provided key, and
+   hen stored on the disk
+
+### Implementation Details
+
+#### The encryption key
+
+The encryption key will use the same format as Bootstrap Tokens.
+See [Authenticating with Bootstrap Tokens](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#token-format)
+for more details.
+
+The default lifecycle of the encryption key will be the following:
+
+- a random encryption-key will be created by `kubeadm init --upload-certs`
+- the encryption-key will be printed in the output of the `kubeadm init --upload-certs`
+  command, but never stored in the cluster
+- the encryption-key will expire/become useless as soon as the `kubeadm-certs` will
+  be automatically deleted
```

Comment:
> the encryption-key will expire/become useless as soon as the `kubeadm-certs` will be automatically deleted

That's not really accurate... if the content of the kubeadm-certs was already obtained, the encryption-key could later decode it. That's why the bootstrap token length is ok for a credential we can revoke, but not for encryption of data that can be decoded over a longer period of time.

Reply:
Point taken. I added a note here and also in the risks mitigation chapter where all the potential security concerns are summarized.